chain/chunks/src/lib.rs

```diff
@@ -179,52 +185,77 @@ impl SealsManager {
+                    &self.runtime_adapter.get_epoch_id_from_prev_block(parent_hash)?,
+                    height,
+                    shard_id,
+                )?;
+                let candidates = {
+                    let n = self.runtime_adapter.num_total_parts();
+                    // `n` is an upper bound for elements in the accumulator; declaring with
+                    // this capacity up front will mean no further allocations will occur
+                    // from `push` calls in the loop.
+                    let mut accumulator = Vec::with_capacity(n);
+
+                    for part_ord in 0..n {
+                        let part_ord = part_ord as u64;
+                        let part_owner =
+                            self.runtime_adapter.get_part_owner(parent_hash, part_ord)?;
+                        if part_owner == chunk_producer || Some(part_owner) == self.me {
+                            continue;
+                        }
+                        accumulator.push(part_ord);
+                    }
+
+                    accumulator
+                };
+
+                let chosen = candidates
+                    .choose_multiple(
+                        &mut rand::thread_rng(),
+                        cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
+                    )
+                    .cloned()
+                    .collect::<HashSet<_>>();
+                let seal = Seal { part_ords: chosen, chunk_producer, sent: Utc::now(), height };
+
+                Ok(entry.insert(seal))
             }
-            let chosen = candidates
-                .choose_multiple(
-                    &mut rand::thread_rng(),
-                    cmp::min(NUM_PARTS_REQUESTED_IN_SEAL, candidates.len()),
-                )
-                .cloned()
-                .collect::<HashSet<_>>();
-            Seal { part_ords: chosen, chunk_producer, sent: Utc::now() }
-        }))
+        }
     }
 
     fn approve_chunk(&mut self, chunk_hash: &ChunkHash) {
         let seal = self.seals.get_mut(chunk_hash).expect("seal should be already produced");
         seal.part_ords.clear();
     }
 
-    fn track_seals(&mut self) {
+    fn track_seals(&mut self, block_hash: &CryptoHash) {
         let now = Utc::now();
-        for (chunk_hash, seal) in self.seals.iter_mut() {
-            if seal.part_ords.len() > NUM_PARTS_LEFT_IN_SEAL
-                && (now - seal.sent).num_milliseconds() > ACCEPTING_SEAL_PERIOD_MS
-            {
-                warn!(target: "client", "Couldn't reconstruct chunk {:?} from {:?}, I'm {:?}", chunk_hash, seal.chunk_producer, self.me);
-                self.dont_include_chunks_from.insert(seal.chunk_producer.clone());
+        let gc_stop_height = self.runtime_adapter.get_gc_stop_height(block_hash).ok();
+        let me = &self.me;
+        let dont_include_chunks_from = &mut self.dont_include_chunks_from;
+
+        self.seals.retain(|chunk_hash, seal| {
```

Comment:
This is potentially 200k iterations. I think it is sufficient to keep something like last 1000 heights (similar to what `EncodedChunksCache` does).

Reply:
@Kouprin suggested using the GC height as opposed to introducing a new constant because the GC height has some logic related to epochs which a simple constant would not take into account.

Do you think it would be safe to have an arbitrary cut-off point? Unlike the `EncodedChunksCache` there is no persistent store to recover the elements from once they are gone.

Reply:
I don't see why it needs to be persistent. I think Seal is only relevant at the head of the chain unless there is a massive reorg, which I don't think is practical. Also, losing a seal is not the end of the world :)

Reply:
@SkidanovAlex what do you think?

Reply:
We don't need to iterate thru all Seals every time. Seals with empty `part_ords` are necessary to not produce the Seal with same Chunk Hash again. Those Seals could be moved easily into separate HashSet and not be tracked. And those Seals should be GCed.

The only Seals that we honestly tracking are the Seals with non-empty `part_ords`. They are alive only for period `ACCEPTING_SEAL_PERIOD_MS`. Their number is small enough to track them straightforwardly.

@bowenwang1996 the important thing is not to put good validators into `dont_include_chunks_from` array.

Reply:
Do we need to remember seals with empty `part_ords` at all then? Instead of removing based on height, could we remove if
1. The seal is approved
2. `ACCEPTING_SEAL_PERIOD_MS` passes (in this case we also add the chunk producer to `dont_include_chunks_from`)

Reply:
It's not enough to remove Seal if (1) and (2) because next time when we receive Chunk Part we will incorrectly create a new Seal and challenge Chunk Creator once again. Please have a look into `get_seal`.

Reply:
We can move used Seals to separate HashSet.

Reply:
Ah, right, that makes sense. So we still need to be able to tell when we see some `ChunkHash` if we have already challenged it or not.

We might say there are really two types of objects in play. A "seal challenge" is a non-empty list of requests parts for which `ACCEPTING_SEAL_PERIOD_MS` has not yet elapsed. While a "seal result" either indicates all parts were received or some parts were not before the time elapsed.  A "seal challenge" is meant to be short lived, while a "seal result" must stick around for some time so we know not to challenge a `ChunkHash` again.

I think the question still remains what the lifetime of a "seal result" should be. If we store them with a simple `HashMap` then we are in the same situation where either it grows indefinitely (not acceptable) or there is an O(N) pruning loop (possibly an expensive operation). Could we use a `SizedCache` and agree on a number of seal results we need to remember? Or is that too simplistic and the pruning loop is the only way to go? In which case the question becomes, when should we run the pruning loop?

Reply:
Oh, no, I see what you are saying now with your `HashMap<BlockHeight, ChunkHash>` suggestion. But I think it should really be `BTreeMap<BlockHeight, HashSet<ChunkHash>>`

This way checking whether we have challenged a `ChunkHash` previously or not is O(log(n)), where n is the number of heights we choose to cache, and pruning is also quick because the keys in a `BTreeMap` are ordered, so we just remove from the end if it is too old.

I think that is a really nice solution.

The only question is now how many height back from the top should we keep? Should we still use the GC limit, or a separate constant as @bowenwang1996 suggested?

Reply:
> Oh, no, I see what you are saying now with your HashMap<BlockHeight, ChunkHash> suggestion. But I think it should really be BTreeMap<BlockHeight, HashSet<ChunkHash>>

That's right.

> The only question is now how many height back from the top should we keep?

We need to store only ShardChunk Hashes. I suggest to start with the same threshold as GC uses to be consistent in deleting data + not thinking about another threshold and prove that it's chosen properly. Finally, it's easy to change threshold in future, if necessary.

Reply:
I think doing 1000 iterations here is also not acceptable, given this method is called per chunk part received, it must be way faster than iterating over 1000 elements of a hashmap.

Some more sophisticated approach is needed here.

Reply:
@birchmd I think it works except that I don't think the deletion operation you want is supported by `BTreeMap`.

Reply:
> I think doing 1000 iterations here is also not acceptable

It's limited by amount of fresh Chunks that appeared in last 30 seconds.

> Some more sophisticated approach is needed here.

We can save last tracking time and don't run tracing if previous one was < 1 second ago.

Reply:
@bowenwang1996 Yes, you're right it is not supported directly. I think it would be easy to derive for our case though

```rust
fn prune(...) {
    let gc_height = ...;
    let keys_to_remove = b_tree_map.keys().take_while(|h| h < gc_height).collect();
    for k in keys_to_remove {
      b_tree_map.remove(k);
    }
}
```

Most of the time `keys_to_remove` will contain fewer than two elements (at least, as long as there are not regular large jumps in the `gc_height`) and because it is based on an iterator this translate into very little actual work. So the overall prune operation should be fast.

@SkidanovAlex I'll refactor so that this `track_seals` function only loops over the active challenges as @Kouprin suggested, then it should always be over a relatively small number of elements. And I could also put in the optimization he suggested where `track_seals` is a no-op if invoked a second time, in quick succession after the first. Although I'm not sure if this is really needed, remember that in master right now it loops over all seals that have been created _ever_, and I'm not aware if it being a bottleneck currently.

Reply:
@birchmd Maybe I misunderstood something but this collection (seal result) is not supposed to be super ephemeral according to previous discussion right? Therefore this collection is potentially large and your proposed way to garbage collect old data will potentially be slow. Please note that `gc_stop_height` doesn't increase linearly with block height but changes once every epoch so you are looking at 43k iterations (and more than 200k the first time gc stop height is updated), which seems suboptimal.

By "not supported" what I meant is that `BTreeMap` doesn't support fast deletion of subtrees. This operation can be done in O(logn) time for self-balancing BSTs, although the implementation can be nontrivial.

Reply:
@bowenwang1996 Ah, I see. Yes, then the linear loop would be quite bad, yes. There is a [split_off](https://doc.rust-lang.org/std/collections/btree_map/struct.BTreeMap.html#method.split_off) method on `BTreeMap` which may be useful here, although the docs to not indicate the complexity of this operation.

Reply:
The [PR which introduced split_off](https://github.com/rust-lang/rust/pull/33947) claims it is `O(log n)`, so should work for this purpose.

Reply:
I didn't look close into it but the PR description says
```
It splits the tree with O(log n) operations and then calculates sizes by traversing the lower one.
```
If it were to traverse the lower one it seems that the overall complexity would still be linear.

Reply:
Hmm, good point. Is there a way we could benchmark this and see whether this GC strategy is acceptable?

Reply:
@bowenwang1996 it is `O(left size of the split)`, and my understanding is that we are going to perform that many operations after the split anyway.

Reply:
Updated with the current proposal in https://github.com/nearprotocol/nearcore/pull/2702/commits/50fbd5517842a9dc992750486e170fbedc6e20e3

No operations are needed after the split because we just want to drop everything on the one side. The question is whether the split operation is too expensive. It would be nice to test this empirically, or I could try something else if someone has another idea.

Reply:
@SkidanovAlex as Michael said we are not going to do any more operations after the split. If we were to do O(left size of the split) operations we might as well remove the elements naively.
I still don't see why my idea of maintaining less heights is problematic. This way we can maintain the tail however we want (for example dropping the last one or the last few at every call).

Reply:
I updated to a version with a fixed height horizon in https://github.com/nearprotocol/nearcore/commit/1b1b509156500d7da4d48e9ee4869388d562553e

Let me know what you think.